Add lint for items deprecated in future

aheart opened the pull request:
Resolves #55892

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

RalfJung commented on src/test/ui/deprecation/deprecation-in-future.stderr:
```diff
@@ -0,0 +1,14 @@
+error: use of item deprecated in future version 'deprecated_future': text
+  --> $DIR/deprecation-in-future.rs:17:5
+   |
+LL |     deprecated_future(); //~ ERROR use of item deprecated in future version 'deprecated_future': text
```
This looks like `deprecated_future` is the version number. Could it show the version number as well?

aheart commented on src/test/ui/deprecation/deprecation-in-future.rs:
```diff
@@ -0,0 +1,20 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
```
I've removed the copyright. Should I remove the license paragraph too?

RalfJung replied:
Yes, please do.

aheart commented on src/test/ui/deprecation/deprecation-in-future.rs:
```diff
@@ -0,0 +1,16 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
```
@varkor, should I rename this file to deprecation_in_future-lint.rs or something similar?

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/459489998) ([raw log](https://api.travis-ci.org/v3/job/459489998/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:0ae6e514:start=1543177170638669348,finish=1543177224162988327,duration=53524318979
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-5.0
---

[00:03:31] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:03:31] tidy error: /checkout/src/test/ui/deprecation/deprecation-in-future.rs:13: line longer than 100 chars
[00:03:32] tidy error: /checkout/src/librustc/middle/stability.rs:567: line longer than 100 chars
[00:03:32] tidy error: /checkout/src/librustc/middle/stability.rs:600: line longer than 100 chars
[00:03:33] some tidy checks failed
[00:03:33] 
[00:03:33] 
[00:03:33] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:03:33] 
[00:03:33] 
[00:03:33] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:03:33] Build completed unsuccessfully in 0:01:03
[00:03:33] Build completed unsuccessfully in 0:01:03
[00:03:33] make: *** [tidy] Error 1
[00:03:33] Makefile:79: recipe for target 'tidy' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:05bb385c
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Sun Nov 25 20:24:07 UTC 2018
---
travis_time:end:0ba0e3ee:start=1543177447430241162,finish=1543177447435223790,duration=4982628
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0807f89d
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0c22f1ce
travis_time:start:0c22f1ce
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:04e2d13e
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

aheart commented:
@varkor, @RalfJung,
I've addressed your comments, but I'm struggling with the failing travis job.
I can see examples of lines longer than 100 symbols in other files.

RalfJung commented on src/test/ui/deprecation/deprecation-in-future.rs:
```diff
@@ -0,0 +1,16 @@
+pub fn deprecated_future() {}
+
+fn test() {
+    deprecated_future(); //~ ERROR use of item that will be deprecated in a future version 99.99.99 'deprecated_future': text
```
The item name seems oddly out of place where it is.

What about
```
use of item 'deprecated_future' that will be deprecated in a future version 99.99.99: text
```

aheart replied:
Adjusted

RalfJung commented:
> I can see examples of lines longer than 100 symbols in other files.

You probably have to add `// ignore-tidy-linelength` to your test.

aheart commented:
> You probably have to add // ignore-tidy-linelength to your test.
That explains it. Thank you!

aheart commented on src/librustc/middle/stability.rs:
```diff
@@ -564,18 +564,20 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
+        let lint_deprecated = |def_id: DefId,
+                               id: NodeId,
+                               note: Option<Symbol>,
+                               message: &str,
```
I wonder if it would be more appropriate to not pass the message in at all and instead to choose the message inside the closure by matching against `lint`

varkor replied:
The `Lint` just has a string identifier, rather than an enum, so we wouldn't get any assurances that this code would be updated if the lints ever changed, so I think passing it in is a better choice here.

RalfJung commented on src/librustc/middle/stability.rs:
```diff
@@ -592,11 +594,30 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
+                if deprecated_in_future_version {
+                    let path = self.item_path_str(def_id);
+                    let message = format!("use of item '{}' \
+                                           that will be deprecated in a future version {}",
```
```suggestion
                                           that will be deprecated in future version {}",
```

varkor commented on src/librustc/middle/stability.rs:
```diff
@@ -579,6 +579,21 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             }
         };
 
+        let lint_future_deprecated = |def_id: DefId, id: NodeId, note: Option<Symbol>| {
```
Yeah, I'm always uncomfortable seeing an `unwrap` where avoidable. You could update `deprecated_in_future_version` to store the version as an `Option<&str>` and then match on that instead.

varkor commented:
Sorry for taking a while to get back on this. If you could get rid of the unwrap by adding the `since` information to `deprecated_in_future_version` (https://github.com/rust-lang/rust/pull/56203#discussion_r239259835) and squash your commits, then everything looks great to me! Thanks!

bors commented:
:umbrella: The latest upstream changes (presumably #56502) made this pull request unmergeable. Please resolve the merge conflicts.

aheart commented:
@varkor, no problem. Unwrap is gone. I've also rebased unto master and squashed my commits.
I appreciate your and @RalfJung's help. :smiley:

varkor commented on src/librustc/middle/stability.rs:
```diff
@@ -614,8 +639,14 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             if let Some(id) = id {
+                let path = self.item_path_str(def_id);
+                let message = format!("use of deprecated item '{}'", path);
                 if deprecation_in_effect(&since.as_str()) {
```
Sorry, I didn't notice before, but we need to make the same changes here. This warning is for deprecated items inside the standard library, etc. It should be affected by `DEPRECATED_IN_FUTURE` as well.

varkor commented:
Thanks! Just one final comment (which is particularly important for addressing https://github.com/rust-lang/rust/issues/55892).

aheart commented on src/librustc/middle/stability.rs:
```diff
@@ -584,17 +586,40 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
-                if !skip {
-                    lint_deprecated(def_id, id, depr_entry.attr.note);
+
+
```
Fixed

aheart commented:
@varkor I've amended 5c79eab, hopefully I've not misinterpreted "previous commit".

varkor commented:
@aheart: thanks for your work on this!

@bors r+ rollup

bors commented:
:pushpin: Commit f8c03b6ab771f0a83f354d645ff357af83aa29ea has been approved by `varkor`

<!-- @bors r=varkor f8c03b6ab771f0a83f354d645ff357af83aa29ea -->

aheart commented:
@varkor it was a pleasure. :smile:

pietroalbini commented:
@bors rollup-

bors commented:
:hourglass: Testing commit f8c03b6ab771f0a83f354d645ff357af83aa29ea with merge 93c2f055ba443385b7c97b1b81fb179db5c9e065...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/21037270), [status-travis](https://travis-ci.org/rust-lang/rust/builds/468507370?utm_source=github_status&utm_medium=notification)
Approved by: varkor
Pushing 93c2f055ba443385b7c97b1b81fb179db5c9e065 to master...
